app/frontend/src/features/communities/hooks.ts

```diff
@@ -160,6 +167,78 @@ export const useNewDiscussionMutation = (onSuccess?: () => void) => {
   );
 };
 
+export const useListAvailableReferences = (
+  userId: number,
+  type: "received" | "given" | "all"
+) =>
+  useQuery<AvailableWriteReferencesRes.AsObject, GrpcError>(
+    referencesKey(userId, type),
+    () =>
+      service.references.getAvailableReferences({
+        userId,
+      })
+  );
+
+interface WriteHostRequestReferenceVariables {
+  referenceData: WriteHostRequestReferenceInput;
+  setMutationError: SetMutationError;
+}
+
+export function useWriteHostReference(userId: number) {
+  const queryClient = useQueryClient();
+  const { mutate: writeHostRequestReference, status, reset } = useMutation<
+    Reference.AsObject,
+    Error,
+    WriteHostRequestReferenceVariables
+  >(
+    ({ referenceData }) =>
+      service.references.writeHostRequestReference(referenceData),
+    {
+      onError: (error, { setMutationError }) => {
+        setMutationError(error.message);
+      },
+      onMutate: async ({ setMutationError }) => {
+        setMutationError(null);
+      },
+      onSuccess: () => {
+        queryClient.invalidateQueries([referencesKey(userId, "all")]);
```

Comment:
I think we'll need to invalidate the specific user's cache under `referencesReceivedKey` too since this reference may also affect the user's received references

Reply:
```suggestion
        queryClient.invalidateQueries(availableWriteReferencesKey(userId));
```
with the above suggestion on changing the query key to be used and also invalidating the `referencesReceivedKey` as well